### PR TITLE
MGMT-9344: assisted-installer-agent subsystem tests

### PR DESCRIPTION
**Pull request by adriengentil**
Add a new workflow to run agent subsystem tests on an equinix metal
machine. The workflow is composed of the following steps:
- Provision an equinix metal machine (baremetalds step)
- Install required packages to run the tests (skipper and docker) and
  copy agent source code onto the remote machine
- Run the subsystem tests on the remote machine
- Collect the logs and the junit report to the artifact directory
- Teardown the machine (baremetalds step)

The workflow will be run on presubmit and daily for master and
release-ocm-2.4 branches.

### CODE REVIEW

File: ci-operator/step-registry/assisted/agent/test/assisted-agent-test-ref.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+ref:
+  as: assisted-agent-test
+  from: src
```
Comment (osherdp) [inline ci-operator/step-registry/assisted/agent/test/assisted-agent-test-ref.yaml:3]: usually if the step is specific to some code, then it should be using the specific image
e.g.:
```suggestion
  from: assisted-installer-agent-code
```

And in ``openshift-assisted-installer-agent-master.yaml``:
```
images:
...
- from: src
  to: assisted-installer-agent-code
```

Comment (osherdp) [reply ci-operator/step-registry/assisted/agent/test/assisted-agent-test-ref.yaml:3]: same goes to the other steps

**Review by osherdp — CHANGES_REQUESTED** [review]

File: ci-operator/jobs/openshift/assisted-installer-agent/openshift-assisted-installer-agent-master-presubmits.yaml [hunk 2]
```diff
@@ -76,7 +76,7 @@ presubmits:
           secretName: result-aggregator
     trigger: (?m)^/test( | .* )e2e-metal-assisted,?($|\s.*)
   - agent: kubernetes
-    always_run: false
+    always_run: true
```
[the remark this answers is not in the document]
Comment (adriengentil) [reply ci-operator/jobs/openshift/assisted-installer-agent/openshift-assisted-installer-agent-master-presubmits.yaml:79]: yes, it comes from the `make update`, I need to revert these changes.

[the remark this answers is not in the document]
Comment (adriengentil) [reply ci-operator/jobs/openshift/assisted-installer-agent/openshift-assisted-installer-agent-master-presubmits.yaml:79]: yes, it comes from the `make update`, I need to revert these changes.

File: ci-operator/step-registry/assisted/agent/test/assisted-agent-test-ref.yaml (hunk 1 above)
Comment (adriengentil) [reply ci-operator/step-registry/assisted/agent/test/assisted-agent-test-ref.yaml:3]: is there a specific reason to have such indirection or it's just a convention?

File: ci-operator/step-registry/assisted/agent/assisted-agent-workflow.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    test:
+      - ref: assisted-agent-test
+    post:
+      - ref: assisted-agent-gather
+      - ref: baremetalds-packet-teardown
+    env:
+      PACKET_PLAN: "c3.small.x86"
+      PACKET_OS: "rocky_8"
+  documentation: |-
+    This workflow executes the substem test suite on a cluster provisioned by running assisted-installer-agent on a packet server.
```
[the remark this answers is not in the document]
Comment (adriengentil) [reply ci-operator/step-registry/assisted/agent/assisted-agent-workflow.yaml:18]: :eagle:

File: ci-operator/step-registry/assisted/agent/test/assisted-agent-test-ref.yaml (hunk 1 above)
Comment (osherdp) [reply ci-operator/step-registry/assisted/agent/test/assisted-agent-test-ref.yaml:3]: the reason for that indirection is to make it runnable on other repos. ci-operator is built to steps being reused anywhere needed even when it's for jobs of other repos

in this specific case, the major advantage is actually explicitly not allow using it for ``src`` that's originating from other repos. the main scenarios of just using ``src`` in steps is e.g. general step to lint code and such thing

Comment (osherdp) [reply ci-operator/step-registry/assisted/agent/test/assisted-agent-test-ref.yaml:3]: for a completely generic image, I think there's ``cli`` as well

**Comment by osherdp** [comment]
@adriengentil can you add some content to the PR description?

File: ci-operator/config/openshift/assisted-installer-agent/openshift-assisted-installer-agent-master.yaml [hunk 4]
```diff
@@ -124,6 +124,10 @@ tests:
   secret:
     mount_path: /tmp/secret
     name: assisted-installer-agent-codecov-token
+- as: subsystem-test
```
[the remark this answers is not in the document]
Comment (adriengentil) [reply ci-operator/config/openshift/assisted-installer-agent/openshift-assisted-installer-agent-master.yaml]: and postsubmit as well?

File: ci-operator/config/openshift/assisted-installer-agent/openshift-assisted-installer-agent-master.yaml [hunk 5]
```diff
@@ -124,6 +124,10 @@ tests:
...
+  steps:
+    cluster_profile: packet-assisted
```
[the remark this answers is not in the document]
Comment (adriengentil) [reply ci-operator/config/openshift/assisted-installer-agent/openshift-assisted-installer-agent-master.yaml]: Nope, the job fails to start if I remove it from the config

File: ci-operator/step-registry/assisted/agent/test/assisted-agent-test-ref.yaml (hunk 1 above)
Comment (adriengentil) [reply ci-operator/step-registry/assisted/agent/test/assisted-agent-test-ref.yaml]: And also, it makes sense, thanks for the explanation

File: ci-operator/jobs/openshift/assisted-installer-agent/openshift-assisted-installer-agent-master-periodics.yaml [hunk 6]
```diff
@@ -53,3 +53,57 @@ periodics:
...
+    skip_cloning: true
+  extra_refs:
+  - base_ref: master
+    org: openshift
+    repo: assisted-installer-agent
+  labels:
+    ci.openshift.io/generator: prowgen
+    pj-rehearse.openshift.io/can-be-rehearsed: "true"
+  name: periodic-ci-openshift-assisted-installer-agent-master-subsystem-test-nightly
+  spec:
```
[the remark this answers is not in the document]
Comment (osherdp) [reply ci-operator/jobs/openshift/assisted-installer-agent/openshift-assisted-installer-agent-master-periodics.yaml:72]: this one seems unresolved

**Review by osherdp — APPROVED** [review]

**Comment by osherdp** [comment]
Probably need to backport removal of ``Image availability tests Small timeout`` from release-ocm-2.4 branch
But other than that, LGTM

**Comment by adriengentil** [comment]
> Probably need to backport removal of `Image availability tests Small timeout` from release-ocm-2.4 branch But other than that, LGTM

It's here: https://github.com/openshift/assisted-installer-agent/pull/309

**Comment by osherdp** [comment]
/test pj-rehearse
seems like infra problems

**Review by osherdp — APPROVED** [review]

**Comment by osherdp** [comment]
/hold
(to wait for just one subsystem job success)

**Comment by adriengentil** [comment]
/retest
For some reasons, the tests failed to provision a packet instance

**Comment by osherdp** [comment]
@adriengentil looks good, no?

**Comment by adriengentil** [comment]
Yes, the subsystem tests are all passing

**Comment by osherdp** [comment]
so we can unhold?

**Pull request merged**